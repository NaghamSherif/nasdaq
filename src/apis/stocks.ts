import axios from "axios";

import { API_URL } from "../constants/apis";
import { TOKEN } from "../constants/apis";

export async function getStocks(pageParam: any, searchQuery: string) {
  const headers = { Authorization: `Bearer ${TOKEN}` };
  return await axios
    .get(pageParam ? pageParam : `${API_URL}/v3/reference/tickers`, {
      headers,
      ...(searchQuery ? { params: { search: searchQuery } } : {}),
    })
    .then((res) => {
      return res.data;
    });
}
