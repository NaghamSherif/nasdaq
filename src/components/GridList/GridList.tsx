import { StyleSheet, View, FlatList } from "react-native";

export default function GridList({
  data,
  numColumns,
  renderItem,
  onEndReached = () => {},
}: {
  data: any[];
  numColumns: number;
  renderItem: (
    item: any,
    margin: { marginRight?: number; marginLeft?: number }
  ) => React.JSX.Element;
  onEndReached?: (info: { distanceFromEnd: number }) => void;
}) {
  const formatData = (data: any[], numColumns: number) => {
    const numberOfFullRows = Math.floor(data.length / numColumns);

    let numberOfElementsLastRow = data.length - numberOfFullRows * numColumns;
    while (
      numberOfElementsLastRow !== numColumns &&
      numberOfElementsLastRow !== 0
    ) {
      data.push({ key: `blank-${numberOfElementsLastRow}`, empty: true });
      numberOfElementsLastRow++;
    }

    return data;
  };

  const renderFunc = ({ item, index }: { item: any; index: number }) => {
    if (item.empty === true) {
      return <View style={styles.itemInvisible} />;
    }
    return renderItem(
      item,
      index % 2 === 0
        ? {
            marginRight: 10,
          }
        : {
            marginLeft: 10,
          }
    );
  };

  return (
    <FlatList
      data={formatData(data, numColumns)}
      style={styles.container}
      renderItem={renderFunc}
      numColumns={numColumns}
      showsHorizontalScrollIndicator={false}
      showsVerticalScrollIndicator={false}
      onEndReached={(e) => {
        if (e?.distanceFromEnd === 0) onEndReached(e);
      }}
      onEndReachedThreshold={0.001}
    />
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: "100%",
    margin: 0,
    marginTop: 10,
    marginBottom: 0,
    gap: 10,
  },
  itemInvisible: {
    flex: 1,
    backgroundColor: "transparent",
    borderRadius: 18,
    marginBottom: 20,
  },
});
