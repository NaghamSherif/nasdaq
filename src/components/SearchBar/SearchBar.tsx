import { useState } from "react";
import { StyleSheet, TextInput } from "react-native";

import colors from "../../constants/colors";

export default function SearchBar({
  onSubmit,
}: {
  onSubmit: (text: string) => void;
}) {
  const [value, onChange] = useState("");
  return (
    <TextInput
      style={styles.searchBar}
      onChangeText={onChange}
      onSubmitEditing={(e) => {
        onSubmit(e.nativeEvent.text);
      }}
      value={value}
      placeholder="Search for stocks"
      placeholderTextColor={colors?.grayText}
      testID="input"
    />
  );
}

const styles = StyleSheet.create({
  searchBar: {
    width: "100%",
    borderRadius: 100,
    borderWidth: 1,
    borderStyle: "solid",
    borderColor: colors?.cardBorder,
    backgroundColor: colors?.navy,
    padding: 12,
    paddingStart: 16,
    color: colors?.grayText,
  },
});
