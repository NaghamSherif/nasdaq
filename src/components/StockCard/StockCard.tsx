import { StyleSheet, Text, View } from "react-native";

import colors from "../../constants/colors";

export default function StockCard({
  ticker,
  name,
  margin,
}: {
  ticker: string;
  name: string;
  margin: { marginRight?: number; marginLeft?: number };
}) {
  return (
    <View style={[styles.container, margin]}>
      <View style={styles?.logo}>
        <Text style={styles?.logoAlt}>AA</Text>
      </View>
      <Text style={styles?.ticker}>{ticker}</Text>
      <Text numberOfLines={1} style={styles?.name}>
        {name}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors?.navy,
    alignItems: "center",
    justifyContent: "center",
    borderRadius: 18,
    marginBottom: 20,
    aspectRatio: 1,
    borderWidth: 1,
    borderStyle: "solid",
    borderColor: colors?.cardBorder,
  },
  logo: {
    height: 35,
    width: 35,
    alignItems: "center",
    justifyContent: "center",
    borderRadius: 8,
    borderWidth: 1,
    borderStyle: "solid",
    borderColor: colors?.cardBorder,
    padding: 5,
    marginBottom: 20,
  },
  logoAlt: {
    color: colors?.grayText,
    fontSize: 14,
  },
  ticker: {
    fontWeight: "600",
    fontSize: 16,
    color: colors?.lightText,
    marginBottom: 8,
  },
  name: {
    fontSize: 14,
    color: colors?.grayText,
    textAlign: "center",
    maxWidth: "80%",
    overflow: "hidden",
  },
});
