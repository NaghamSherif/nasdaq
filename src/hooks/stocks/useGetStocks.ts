import { useInfiniteQuery } from "@tanstack/react-query";
import { getStocks } from "../../apis/stocks";

export function useGetStocks(searchQuery: string) {
  return useInfiniteQuery({
    queryKey: ["stocks-list", searchQuery],
    queryFn: ({ pageParam = 0 }) => getStocks(pageParam, searchQuery),
    initialPageParam: 0,
    getNextPageParam: (lastPage) => {
      return lastPage.next_url;
    },
  });
}
