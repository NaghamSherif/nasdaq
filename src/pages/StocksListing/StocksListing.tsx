import { ActivityIndicator, StyleSheet, View } from "react-native";
import { useEffect, useState } from "react";

import { useGetStocks } from "../../hooks/stocks";

import StockCard from "../../components/StockCard";
import GridList from "../../components/GridList";
import SearchBar from "../../components/SearchBar";
import colors from "../../constants/colors";

export default function StocksListing() {
  const [stocks, setStocks] = useState<any[]>([]);
  const [searchQuery, setSearchQuery] = useState("");
  const { data, isLoading, isFetching, isSuccess, hasNextPage, fetchNextPage } =
    useGetStocks(searchQuery);

  useEffect(() => {
    isSuccess &&
      setStocks((prev) => [
        ...prev,
        ...data?.pages?.[data?.pages?.length - 1]?.results,
      ]);
  }, [data?.pages?.length]);

  useEffect(() => {
    setStocks([]);
  }, [searchQuery]);

  const loadNextPageData = () => {
    if (hasNextPage) {
      fetchNextPage();
    }
  };

  return (
    <View style={styles.container}>
      <View style={styles.searchBarContainer}>
        <SearchBar onSubmit={setSearchQuery} />
      </View>
      {(isLoading || isFetching) && !data?.pages?.[0]?.results?.length ? (
        <ActivityIndicator
          size="large"
          color={colors?.white}
          style={styles.commonContainer}
        />
      ) : isSuccess ? (
        <View style={styles.commonContainer}>
          <GridList
            data={stocks}
            numColumns={2}
            renderItem={(item, margin) => (
              <StockCard
                ticker={item.ticker}
                name={item.name}
                margin={margin}
              />
            )}
            onEndReached={loadNextPageData}
          />
          {isLoading || isFetching ? (
            <ActivityIndicator
              size="large"
              color={colors?.white}
              style={{ paddingBottom: 12 }}
            />
          ) : null}
        </View>
      ) : null}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    width: "100%",
    backgroundColor: colors?.appBackground,
    alignItems: "center",
    justifyContent: "flex-start",
    padding: 18,
    paddingTop: 16,
    paddingBottom: 0,
  },
  searchBarContainer: {
    width: "100%",
    marginBottom: 8,
  },
  commonContainer: {
    flex: 1,
    width: "100%",
    height: "100%",
  },
});
