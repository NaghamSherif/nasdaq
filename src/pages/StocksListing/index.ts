import StocksListing from "./StocksListing";

export default StocksListing;
