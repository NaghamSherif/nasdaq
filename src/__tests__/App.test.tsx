import React from "react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

import {
  render,
  screen,
  waitFor,
  act,
  fireEvent,
} from "@testing-library/react-native";

import GridList from "../components/GridList";
import StockCard from "../components/StockCard";
import StocksListing from "../pages/StocksListing";
import SearchBar from "../components/SearchBar";

import { useGetStocks } from "../hooks/stocks";

jest.mock("../hooks/stocks", () => ({
  useGetStocks: jest.fn(),
}));

const queryClient = new QueryClient();

describe("App test Suite", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("GridList renders the correct data", async () => {
    render(
      <GridList
        data={[
          { ticker: "AAPL", name: "Apple Inc." },
          { ticker: "GOOGL", name: "Alphabet Inc." },
        ]}
        numColumns={2}
        renderItem={(item, margin) => (
          <StockCard ticker={item.ticker} name={item.name} margin={margin} />
        )}
      />
    );
    expect(screen.getByText("AAPL")).toBeTruthy();
  });

  it("renders a loader", async () => {
    (useGetStocks as jest.Mock).mockImplementation(() => ({
      isLoading: true,
    }));

    const { getByTestId } = render(
      <QueryClientProvider client={queryClient}>
        <StocksListing />
      </QueryClientProvider>
    );
    await waitFor(() => expect(getByTestId("loader")).toBeTruthy());
  });

  it("renders a list of stocks", async () => {
    const { getByText, rerender } = render(
      <QueryClientProvider client={queryClient}>
        <StocksListing />
      </QueryClientProvider>
    );

    (useGetStocks as jest.Mock).mockImplementation(() => ({
      isSuccess: true,
      data: {
        pages: [
          {
            results: [
              { ticker: "AAPL", name: "Apple Inc." },
              { ticker: "GOOGL", name: "Alphabet Inc." },
            ],
          },
        ],
      },
    }));

    rerender(
      <QueryClientProvider client={queryClient}>
        <StocksListing />
      </QueryClientProvider>
    );

    await act(async () => {
      await waitFor(() => expect(getByText("AAPL")).toBeTruthy());
    });
  });

  it("Search bar works correctly", async () => {
    let mockSubmit = jest.fn();
    const { getByTestId } = render(<SearchBar onSubmit={mockSubmit} />);

    const input = getByTestId("input");
    fireEvent.changeText(input, "micro");
    fireEvent(input, "submitEditing", { nativeEvent: { text: "" } });

    expect(mockSubmit).toHaveBeenCalled();
  });
});
