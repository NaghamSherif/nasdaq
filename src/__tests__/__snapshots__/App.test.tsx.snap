// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#191a28",
      "flex": 1,
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#191a28",
        "flexDirection": "row",
        "height": 55,
        "justifyContent": "flex-start",
        "padding": 10,
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="contain"
      source={1}
      style={
        {
          "height": 40,
          "width": 120,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1f202f",
        "flex": 1,
        "justifyContent": "flex-start",
        "padding": 18,
        "paddingBottom": 0,
        "paddingTop": 16,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "marginBottom": 8,
          "width": "100%",
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search for stocks"
        placeholderTextColor="#a4a4ad"
        style={
          {
            "backgroundColor": "#232639",
            "borderColor": "#2e3040",
            "borderRadius": 100,
            "borderStyle": "solid",
            "borderWidth": 1,
            "color": "#a4a4ad",
            "padding": 12,
            "paddingStart": 16,
            "width": "100%",
          }
        }
        value=""
      />
    </View>
    <ActivityIndicator
      color="#fff"
      size="large"
      style={
        {
          "flex": 1,
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </View>
</RCTSafeAreaView>
`;
